File: avientek/public/js/expense_claim.js
frappe.ui.form.on('Expense Claim', {
    refresh: function(frm) {
        if (frm.doc.docstatus === 1) {
            frm.add_custom_button("Payment Request Form", function() {
                frappe.model.open_mapped_doc({
                    method: "avientek.events.expense_claim.create_payment_request",
                    frm: frm
                });
            }, "Create");
        }
    },

    // ── Client Script: "Company Filter based on employee" ──
    employee: function(frm) {
        if (frm.doc.employee) {
            frappe.call({
                method: 'frappe.client.get_value',
                args: {
                    doctype: 'Employee',
                    filters: { name: frm.doc.employee },
                    fieldname: ['company']
                },
                callback: function(r) {
                    if (r.message) {
                        frm.set_query('company', function() {
                            return {
                                filters: { 'name': r.message.company }
                            };
                        });
                    }
                }
            });
        }
    }
});

File: avientek/events/expense_claim.py
import frappe
from frappe.model.mapper import get_mapped_doc
from frappe.utils import flt

@frappe.whitelist()
def create_payment_request(source_name, target_doc=None, args=None):
    def set_missing_values(source, target):
        target.party_type = "Employee"
        target.party = source.employee
        target.party_name = source.employee_name

        outstanding = flt(source.total_sanctioned_amount) - flt(source.total_amount_reimbursed)

        # Sammish 2026-05-21 (Jithin escalation): Create → Payment Request
        # Form on an Expense Claim was putting the Frappe EC doc name into
        # `reference_name` and leaving `document_reference` empty. That
        # broke the print + Combined PDF resolver. Canonical contract:
        #   - reference_name      = "" (EC has no third-party bill_no)
        #   - document_reference  = Frappe doc name of the EC (canonical
        #                           pointer used by every downstream
        #                           renderer / resolver)
        target.append("payment_references", {
            "reference_doctype": "Expense Claim",
            "reference_name": "",
            "bill_no": "",
            "grand_total": flt(source.total_sanctioned_amount),
            "base_grand_total": flt(source.total_sanctioned_amount),
            "outstanding_amount": outstanding,
            "base_outstanding_amount": outstanding,
            "invoice_date": source.posting_date,
            "document_reference": source.name,
            "currency": frappe.get_cached_value("Company", source.company, "default_currency"),
            "exchange_rate": 1,
        })

        target.total_outstanding_amount = outstanding
        target.total_payment_amount = 0
        target.total_amount = flt(source.total_sanctioned_amount)

    target_doc = get_mapped_doc(
        "Expense Claim",
        source_name,
        {
            "Expense Claim": {
                "doctype": "Payment Request Form",
                "field_map": {
                    "company": "company",
                    "department": "department",
                    "cost_center": "cost_center",
                },
            },
        },
        target_doc,
        postprocess=set_missing_values,
    )

    return target_doc
